pkg/logcli/client/client_test.go

```diff
@@ -78,3 +91,198 @@ func Test_getHTTPRequestHeader(t *testing.T) {
+	then := now.Add(-1 * time.Hour)
+
+	server := NewMockLokiHTTPServer(t, now, then)
+	client := &DefaultClient{
+		Address: fmt.Sprintf("http://127.0.0.1%s", server.server.Addr),
+	}
+
+	server.Run(t)
+	time.Sleep(time.Second)
+	defer server.Stop(t)
+
+	t.Run("GetVolume happy path", func(t *testing.T) {
+		query := volume.Query{
+			QueryString:       `{foo="bar"}`,
+			Start:             then,
+			End:               now,
+			Step:              10 * time.Minute,
+			Quiet:             false,
+			Limit:             10,
+			TargetLabels:      []string{},
+			AggregateByLabels: false,
+		}
+		resp, err := client.GetVolume(&query)
+		require.NoError(t, err)
+
+		require.Equal(t, loghttp.QueryResponse{
+			Status: loghttp.QueryStatusSuccess,
+			Data: loghttp.QueryResponseData{
+				ResultType: loghttp.ResultTypeVector,
+				Result: loghttp.Vector{
+					{
+						Metric: map[model.LabelName]model.LabelValue{
+							"foo": "bar",
+						},
+						Value:     42,
+						Timestamp: model.TimeFromUnixNano(now.UnixNano()),
+					},
+				},
+			},
+		}, *resp)
+	})
+
+	t.Run("GetVolumeRange happy path", func(t *testing.T) {
+		query := volume.Query{
+			QueryString:       `{foo="bar"}`,
+			Start:             then,
+			End:               now,
+			Step:              10 * time.Minute,
+			Quiet:             false,
+			Limit:             10,
+			TargetLabels:      []string{},
+			AggregateByLabels: false,
+		}
+		resp, err := client.GetVolumeRange(&query)
+		require.NoError(t, err)
+
+		require.Equal(t, loghttp.QueryResponse{
+			Status: loghttp.QueryStatusSuccess,
+			Data: loghttp.QueryResponseData{
+				ResultType: loghttp.ResultTypeMatrix,
+				Result: loghttp.Matrix{
+					{
+						Metric: map[model.LabelName]model.LabelValue{"foo": "bar"},
+						Values: []model.SamplePair{
+							{
+								Timestamp: model.TimeFromUnixNano(then.UnixNano()),
+								Value:     42,
+							},
+							{
+								Timestamp: model.TimeFromUnixNano(now.UnixNano()),
+								Value:     47,
+							},
+						},
+					},
+				},
+			},
+		}, *resp)
+
+	})
+
+}
+
+type mockLokiHTTPServer struct {
+	server   *http.Server
+	tenantID string
+	now      time.Time
+	then     time.Time
+}
+
+func NewMockLokiHTTPServer(t *testing.T, now, then time.Time) *mockLokiHTTPServer {
+	listener, err := net.Listen("tcp", ":0")
+	require.NoError(t, err)
+
+	port := listener.Addr().(*net.TCPAddr).Port
+	err = listener.Close()
+	require.NoError(t, err)
+
+	return &mockLokiHTTPServer{
+		server: &http.Server{Addr: fmt.Sprintf(":%d", port), Handler: nil},
+		now:    now,
+		then:   then,
+	}
+}
+
+func (s *mockLokiHTTPServer) getTenantIDUnsafe() string {
+	return s.tenantID
+}
+
+func (s *mockLokiHTTPServer) Run(t *testing.T) {
+	var mux http.ServeMux
+	labels := labels.Labels{
+		{
+			Name:  "foo",
+			Value: "bar",
+		},
+	}
+	codec := queryrange.Codec{}
+
+	mux.HandleFunc("/loki/api/v1/index/volume", func(w http.ResponseWriter, request *http.Request) {
+		volume := queryrange.LokiPromResponse{
```

Comment:
Very detailed test case, thanks your help. My only doubt here is the return value type of this test.

The return value type here should be logproto.VolumeResponse instead of  queryrange.LokiPromResponse, because the code of http.go return logproto.VolumeResponse, and the test construction in http_test.go is also logproto.VolumeResponse

https://github.com/grafana/loki/blob/main/pkg/querier/http.go#L489
```go
        if resp == nil { // Some stores don't implement this
		resp = &logproto.VolumeResponse{Volumes: []logproto.Volume{}}
	}
	if err := queryrange.WriteResponse(r, nil, resp, w); err != nil {
		serverutil.WriteError(err, w)
	}
```

https://github.com/grafana/loki/blob/main/pkg/querier/http_test.go#L386
```go
	ret := &logproto.VolumeResponse{
		Volumes: []logproto.Volume{
			{Name: `{foo="bar"}`, Volume: 38},
		},
	}
```

https://github.com/grafana/loki/blob/main/pkg/util/marshal/marshal.go#L132
```go
func WriteVolumeResponseJSON(r *logproto.VolumeResponse, w io.Writer) error {
	s := jsoniter.ConfigFastest.BorrowStream(w)
	defer jsoniter.ConfigFastest.ReturnStream(s)
	s.WriteVal(r)
	s.WriteRaw("\n")
	return s.Flush()
}
```